harness/determined/core/_state_manager.py

```diff
@@ -0,0 +1,162 @@
+import atexit
+import logging
+import sys
+import threading
+import time
+from types import TracebackType
+from typing import Any, Optional, Union
+
+from determined.common import api
+from determined.common.api import bindings
+
+logger = logging.getLogger("determined.core")
+
+
+class StateManager:
+    """StateManager controls the userspace trial state updates.
+
+    For unmanaged / detached trials, this class will report the state changes
+    to the determinedmaster.
+
+    For managed trials, it'll not do anything as the state is controlled by
+    determined-master itself.
+    """
+
+    def __init__(self, *, session: api.Session, trial_id: int) -> None:
+        self._session = session
+        self._trial_id = trial_id
+
+    def start(self) -> "StateManager":
+        return self
+
+    def close(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> "StateManager":
+        return self
+
+    def __enter__(self) -> "StateManager":
+        return self.start()
+
+    def __exit__(
+        self,
+        exc_type: Optional[type] = None,
+        exc_val: Optional[BaseException] = None,
+        exc_tb: Optional[TracebackType] = None,
+    ) -> "StateManager":
+        return self.close(exc_type, exc_val, exc_tb)
+
+
+class ManagedTrialStateManager(StateManager):
+    """
+    ManagedTrialStateManager leaves the state management to the server.
+    """
+
+    pass
+
+
+class _HeartbeatReporter(threading.Thread):
+    def __init__(self, session: api.Session, trial_id: int) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._should_quit = False
+
+        super().__init__(daemon=True)
+
+    def _post_heartbeat(self) -> None:
+        body = bindings.v1PatchTrialRequest(trialId=self._trial_id)
+        bindings.patch_PatchTrial(session=self._session, body=body, trialId=self._trial_id)
+
+    def run(self) -> None:
+        while not self._should_quit:
+            try:
+                self._post_heartbeat()
+            except Exception:
+                logger.warning(
+                    "failure communicating with heartbeat API (retrying in 10s):", exc_info=True
+                )
+                time.sleep(10)
+            else:
+                time.sleep(60)
+
+    def close(self) -> None:
+        self._should_quit = True
+
+    def __enter__(self) -> "_HeartbeatReporter":
+        self.start()
+        return self
+
+    def __exit__(self, *_: Any) -> None:
+        self.close()
+
+
+class _ExitHook(object):
+    def __init__(self) -> None:
+        self.exit_code = None  # typing: Optional[Union[int, str]]
+        self.exception = None
+
+    def hook(self) -> None:
+        self._orig_exit = sys.exit
+        self._orig_excepthook = sys.excepthook
+        sys.exit = self.exit  # type: ignore
+        sys.excepthook = self.exc_handler
```

Comment:
wut?

This looks terrifying, but I suppose that's what excepthook is there for.  TIL.

Though sys.exit() raises a stop exception, so I'd rather you caught that and captured in the excepthook than modifying `sys.exit()`, which does not seem to be indtended.

Reply:
yeah, also I don't think this would cover the builtin `exit()` as-is, either.  But the way I proposed should cover that, I think.

Reply:
>Though sys.exit() raises a stop exception, so I'd rather you caught that and captured in the excepthook than modifying sys.exit(), which does not seem to be intended.

1. surprisingly, `sys.exit()` call is not handled by `sys.excepthook`. so it seems we have to do it.
2. other chunks of code out there also tend to explicitly handle it.

Reply:
Unfortunate, but ok.

What about the `exit()` builtin?

Reply:
`exit()` and `os._exit` will not be captured. the latter we can't really do anything about. `exit()` is not supposed to be used in the interactive programs, do you still think we need to overwrite it or can we leave it to the general heartbeat mechanism?

Reply:
> os._exit will not be captured

Well I should hope not, since it's just a wrapper around the C `exit()` syscall, I think.

> exit() is not supposed to be used in the interactive programs

This is news to me.

> do you still think we need to overwrite it

I'm not sure if you even can?  Can you?

Reply:
>This is news to me.
"They are useful for the interactive interpreter shell and should not be used in programs."
https://docs.python.org/3/library/constants.html#constants-added-by-the-site-module

>I'm not sure if you even can? Can you?

yeah.
```
import builtins
builtins.exit = lambda: print("hello")
```

Reply:
Whoa. I didn't know they were special objects that had an effect when printed, and I didn't know that they could be overridden.  TIL, TIL.